### PR TITLE
'update' button for bathroom info

### PR DESCRIPTION
**Pull request by Tangbodi**
I added 'update' buttons for every bathroom
Screen Shot 2024-02-26 at 2 01 15 PM

### CODE REVIEW

**Comment by Tangbodi** [comment]
Integrated 'update' viewmodel and put it outside of LazyColumn block, so that the viewmodel won't run repeatedly as many time as the size of listOfBathroomData
Screen Shot 2024-02-26 at 2 35 28 PM

**Comment by SamIAmHarris** [comment]
I think you may have accidentally added a BathroomsPR submodule to project

File: app/src/main/java/com/example/firstandroid/MainActivity.kt [hunk 1]
```diff
@@ -37,6 +43,7 @@ class MainActivity : ComponentActivity() {
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
```
Comment (SamIAmHarris) [inline app/src/main/java/com/example/firstandroid/MainActivity.kt:44]: If we are going to have multiple viewmodels in a class then I would give each a specific name

File: app/src/main/java/com/example/firstandroid/ViewModel/UpdateBathroom/UpdateBathroomData.kt [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+import androidx.lifecycle.ViewModel
+import com.example.firstandroid.BuildConfig.BuildConfig
+import com.example.firstandroid.ViewModel.NewBathroom.NewBathroomData
+import com.example.firstandroid.data.networking.ApiService
+import retrofit2.Retrofit
+import retrofit2.converter.gson.GsonConverterFactory
+import java.math.BigDecimal
+
+
+class UpdateBathroomViewModel() : ViewModel() {
```
Comment (SamIAmHarris) [inline app/src/main/java/com/example/firstandroid/ViewModel/UpdateBathroom/UpdateBathroomData.kt:15]: Can we remove this comment of put it right on the line it is referring to?

File: app/src/main/java/com/example/firstandroid/ViewModel/UpdateBathroom/UpdateBathroomData.kt [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        private set
+
+    fun updateTitle(title: String) {
+        state = state.copy(title = title)
+    }
+
+    fun updateLocation(location: String) {
+        state = state.copy(location = location)
+    }
+
```
Comment (SamIAmHarris) [inline app/src/main/java/com/example/firstandroid/ViewModel/UpdateBathroom/UpdateBathroomData.kt:28]: Good usage of `?.let`

File: app/src/main/java/com/example/firstandroid/ViewModel/UpdateBathroom/UpdateBathroomData.kt [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    fun updateCapacity(capacity: String) {
+        capacity.toIntOrNull()?.let {
+            state = state.copy(capacity = it)
+        }
+    }
+
+    fun updateFree(free: Boolean) {
+        state = state.copy(free = free)
+    }
+
```
Comment (SamIAmHarris) [inline app/src/main/java/com/example/firstandroid/ViewModel/UpdateBathroom/UpdateBathroomData.kt:38]: Good usage of `?.let`

File: app/src/main/java/com/example/firstandroid/ViewModel/UpdateBathroom/UpdateBathroomData.kt [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    fun updateCost(cost: String) {
+        cost.toBigDecimalOrNull()?.let {
+            state = state.copy(cost = it)
+        }
+    }
+
+    fun updateHours(hours: String) {
+        state = state.copy(hours = hours)
+    }
+
```
Comment (SamIAmHarris) [inline app/src/main/java/com/example/firstandroid/ViewModel/UpdateBathroom/UpdateBathroomData.kt:48]: In future we can move this to the repo layer

File: app/src/main/java/com/example/firstandroid/ViewModel/UpdateBathroom/UpdateBathroomData.kt [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    suspend fun updateBathroom(state: UpdateBathroomData) {
+        val apiService: ApiService
+        val retrofit = Retrofit.Builder()
+            .baseUrl(BuildConfig.BASE_URL)
+            .addConverterFactory(GsonConverterFactory.create())
+            .build()
+        apiService = retrofit.create(ApiService::class.java)
+        apiService.updateBathroomData(state)
+    }
+}
```
Comment (SamIAmHarris) [inline app/src/main/java/com/example/firstandroid/ViewModel/UpdateBathroom/UpdateBathroomData.kt:58]: Spacing between the end of the VM and the state object definition please.

File: app/src/main/java/com/example/firstandroid/ViewModel/UpdateBathroom/UpdateBathroomViewModel.kt [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+                    onValueChange = { newHours ->
+                        viewModel.updateHours(newHours)
+                    },
+                    label = { Text("Hours") }
+                )
+            }
+        },
+        confirmButton = {
+            Button(
+                onClick = {
```
Comment (SamIAmHarris) [inline app/src/main/java/com/example/firstandroid/ViewModel/UpdateBathroom/UpdateBathroomViewModel.kt:86]: I don't think we need this coroutine scope here. The ViewModel would be able to do the async work.

File: app/src/main/java/com/example/firstandroid/data/model/BathroomData.kt [hunk 8]
```diff
@@ -13,11 +13,14 @@ data class BathroomData(
     val createdAt: String?,
     val modifiedAt: String?,
 )
```
Comment (SamIAmHarris) [inline app/src/main/java/com/example/firstandroid/data/model/BathroomData.kt:15]: Good class names

File: app/src/main/java/com/example/firstandroid/MainActivity.kt [hunk 9]
```diff
@@ -51,7 +58,7 @@ class MainActivity : ComponentActivity() {
                 setContent {
                     val listOfBathroomData = response.body()?.data
                     var isDialogOpen by remember { mutableStateOf(false) }
-
```
Comment (SamIAmHarris) [inline app/src/main/java/com/example/firstandroid/MainActivity.kt:54]: Ways to handle state across compositions:

1. Remember the state in the Composable. The state will get wiped if the app is put in the background and OS needs to reclaim resources.

2. Keep the state in the viewmodel and pass it to the Composable. This state will survive even if the app is destroyed by the OS.

**Review by SamIAmHarris — CHANGES_REQUESTED** [review]
Let's address some of these comments and get rid of the accidental BathroomPR.

Good progress

**Pull request merged**